user directed retries

andreimatei opened the pull request:
<!-- Reviewable:start -->

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/5116)

<!-- Reviewable:end -->

andreimatei committed:
Support user-directed retries through the RETRY INTENT, RELEASE, RESTART
syntax.

fixes #4877, fixes #4370
referencing #2625

andreimatei committed:
move handling of COMMIT and ROLLBACK to executor

They belong better in the executor than in the planner, since they don't
have plan nodes (their execution used to happen at plan creation time).
Also we want them to directly manipulate txnState, which is awkward to
do from inside the planner (for example ROLLBACK needs to swallow errors
but still close the transaction).

We now also allow retrying on commit errors, which didn't work before
because Commit always cleaned up the txn.

Improved a test to check failures on commit and that txns are actually reused on retries.

andreimatei commented:
The 2 commits can be review separately.

---

Review status: 0 of 23 files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-:-KCXffN_ihdnk33b2Gau:-1357163075)*
<!-- Sent from Reviewable.io -->

tamird commented:
First commit message:

> We now also allow retrying on commit errors, which didn't work before
because Commit always cleaned up the txn.

Is this possible to test?

> Improved a test to check failures on commit and that txns are actually reused on retries.

Where? I don't see any test changes in this commit.

---

Reviewed 5 of 5 files at r1, 10 of 20 files at r2.
Review status: 13 of 23 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[client/txn.go, line 455 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCXkwPDfSPVV2aOe7aR:-KCXkwPE4-Af__0O3dZj:1383055798) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/client/txn.go#L455)):*
it does?

---

*[roachpb/errors.go, line 559 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCXlPVX5M08gNR9YoxR:-KCXlPVX5M08gNR9YoxS:1492448194) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/roachpb/errors.go#L559)):*
`e` is unused

---

*[roachpb/errors.go, line 572 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCXlUgkRRKyfxsMYxxG:-KCXlUgkRRKyfxsMYxxH:1417403230) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/roachpb/errors.go#L572)):*
nit: you can avoid repeating yourself here
```
msg := "current transaction is aborted, commands ignored until end of transaction block"
if e.CustomMsg != "" {
  msg += "; " + e.CustomMsg
}
return msg
```

---

*[roachpb/errors.proto, line 222 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCXlpq7-kDhcG9u0F-N:-KCXlpq7-kDhcG9u0F-O:-196456709) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/roachpb/errors.proto#L222)):*
is it a good idea to combine these notions?

---

*[sql/executor.go, line 602 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCXhIUecFH0pCsTB2s4:-KCXhIUfJDrhcJwoYFQG:-794540581) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ddd0104f832467d80279d38f9172a90eda45577/sql/executor.go#L602)):*
you're forcing allocations here - use two constant strings instead.

---

*[sql/executor.go, line 889 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCXjHOteduJ1tzpfOun:-KCXjHOteduJ1tzpfOuo:997690667) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ddd0104f832467d80279d38f9172a90eda45577/sql/executor.go#L889)):*
this comment (and the other one like it) should probably be on the `return e.{commit,rollback}SqlTransaction` line - it's really very confusing here.

---

*[sql/executor.go, line 932 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCXjq-PZrLSqxC4i-Xs:-KCXjq-PZrLSqxC4i-Xt:-317445410) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ddd0104f832467d80279d38f9172a90eda45577/sql/executor.go#L932)):*
this refactor seems really odd to me. you've exchanged a method on the planner for a method on the executor which takes a planner as an argument and doesn't access any of the executor's state. huh?

---

*[sql/session.proto, line 35 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCXm3Q35ZnRYTg8b2RP:-KCXm3Q35ZnRYTg8b2RQ:1806311243) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/session.proto#L35)):*
indentation is all over the place here

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-:-KCXigGhzbnAdWWXxlDv:92247747)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
I can't speak on the PGWire changes or the retry syntax - was there a consensus on the mailing list discussion? Other than that, mostly just minor comments. There is a lot of random state updating going on across the bank. I wish that could be unified somehow.
Also, as a sidenote, it's hard to review such a large change in detail. I understand that some changes are hard to break up, though. Something to consider whenever possible.

---

Reviewed 5 of 5 files at r1, 20 of 20 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 36 unresolved discussions, some commit checks failed.

---

*[client/db.go, line 396 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCaxyeKfI3K2tPDZPYy:-KCaxyeLlDvHbZKtvR5i:-1413478157) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/client/db.go#L396)):*
if `CleanupOnError` were a no-op when called with a nil error, I think its overall usage would be more elegant.

---

*[client/txn.go, line 502 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCavyVVvr4s6aNPuX98:-KCavx9WTrLWpUfAW5v8:-96652662) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ddd0104f832467d80279d38f9172a90eda45577/client/txn.go#L502)):*
only reviewing the first revision now, but looks like this should be gone at the end of the patch sets.

---

*[client/txn.go, line 454 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCayiiYheeKLkxED1Aw:-KCayiiYheeKLkxED1Ax:1794794463) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/client/txn.go#L454)):*
Can you make this more affirmative: Exec does not attempt to clean up the transaction before returning an error.

---

*[client/txn.go, line 455 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCXkwPDfSPVV2aOe7aR:-KCayDdajMKV1J4V37U0:-559672509) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/client/txn.go#L455)):*
(*txnSender).Send above does this implicitly.

---

*[roachpb/errors.proto, line 180 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCayyS3evosstw8khlc:-KCayyS3evosstw8khld:-1140571795) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/roachpb/errors.proto#L180)):*
I thought we talked about this not being idiomatic. These errors are only in the protos because SQL returns `*Error` internally. Not your job to clean this up, but add a fat TODO on this area on the protos please.

---

*[roachpb/errors.proto, line 222 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCXlpq7-kDhcG9u0F-N:-KCazK1AUNq9MOfuhJ6b:-140865903) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/roachpb/errors.proto#L222)):*
They're actually the same case. Reword as

> NONE (the default) is used for errors which have no effect on the transaction state. That is, a transactional operation which receives such an error is still PENDING and does not need to restart (at least not as a result of the error). Examples are a CPut whose condition wasn't met, or a spurious RPC error.

---

*[sql/executor.go, line 892 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCawhdOKR75yUI8g4SQ:-KCawhdOKR75yUI8g4SR:-1768237136) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ddd0104f832467d80279d38f9172a90eda45577/sql/executor.go#L892)):*
is this the right state? I thought this flag marked the "waiting for an explicit ROLLBACK" mode. Or is that what we want?

---

*[sql/executor.go, line 931 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCaxUbixj1ZTHtawUAP:-KCaxUbixj1ZTHtawUAQ:132266980) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ddd0104f832467d80279d38f9172a90eda45577/sql/executor.go#L931)):*
s/Sql/SQL/g

---

*[sql/executor.go, line 932 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCXjq-PZrLSqxC4i-Xs:-KCawxxbQdNVNtZYEFQL:1269044312) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ddd0104f832467d80279d38f9172a90eda45577/sql/executor.go#L932)):*
Agreed, something's off here. Maybe

```go
// callsite:
p.resetTxn()
{commit,rollback}SQLTransaction(txnState)
```

---

*[sql/executor.go, line 411 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb1QfspuwCuULguKRy:-KCb1QfspuwCuULguKRz:1264779941) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/executor.go#L411)):*
what if txnState contained the execOpts?

---

*[sql/executor.go, line 422 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb-mLLWdHS5HcU0vhf:-KCb-mLLWdHS5HcU0vhg:1359687285) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/executor.go#L422)):*
could you slap a small example in here?

---

*[sql/executor.go, line 444 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb04pYKhExdioW_AFG:-KCb04pYKhExdioW_AFH:-1020338827) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/executor.go#L444)):*
Reword so that this sentence is clearly connected to the case of TransactionAborted.

---

*[sql/executor.go, line 446 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb0QZf9qLHRIKIy1fD:-KCb0QZf9qLHRIKIy1fE:-1146641641) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/executor.go#L446)):*
Add a comment/TODO here that well, we just don't actually back off.

---

*[sql/executor.go, line 618 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb15zEsYa0e83BaKFv:-KCb15zEsYa0e83BaKFw:1552038986) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/executor.go#L618)):*
shouldn't it be possible to set `execOpt.AutoRetry = txnState.autoRetry` somewhere below?

---

*[sql/executor.go, line 649 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb1s8OCO8WDysOgAa0:-KCb1s8OCO8WDysOgAa1:-238027918) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/executor.go#L649)):*
I see repeated updates of both `.state` and `.txn`, time for a helper.

---

*[sql/executor.go, line 745 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb2ObtrBbxQhuuol5Q:-KCb2ObtrBbxQhuuol5R:-1237445069) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/executor.go#L745)):*
you don't use `txnState.txn` for anything, consider moving the above assertion to the caller which allows cleaning up the signature here.

---

*[sql/executor.go, line 853 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb2kDdaNt2wzqpJ0JS:-KCb2kDdaNt2wzqpJ0JT:-2145168684) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/executor.go#L853)):*
s/AbortedTxn/RestartingTxn/?

---

*[sql/executor.go, line 921 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb3PZprftmMkuxqUXE:-KCb3PZprftmMkuxqUXF:-1811194915) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/executor.go#L921)):*
comment rot.

---

*[sql/executor.go, line 955 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb3TtXQjkmQXdBmkqg:-KCb3TtXQjkmQXdBmkqh:-200804101) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/executor.go#L955)):*
this definitely wants a helper. It's everywhere.

---

*[sql/executor.go, line 987 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb3uQfdYvrsbuVn2Se:-KCb3uQfdYvrsbuVn2Sf:734733734) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/executor.go#L987)):*
you don't want this in the `RestartTransaction` case as well?

---

*[sql/executor.go, line 1052 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb4H_a3q4PciBdIQkb:-KCb4EpXBMA8YPaDZnG4:-1340337765) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/executor.go#L1052)):*
helper.

---

*[sql/txn_restart_test.go, line 63 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb4vFias54nuOgI2fw:-KCb4vFias54nuOgI2fx:2108303116) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/txn_restart_test.go#L63)):*
s/val/the val/

---

*[sql/txn_restart_test.go, line 133 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb5Xyqbi3TplyP06T-:-KCb5Xyqbi3TplyP06T0:1487161612) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/txn_restart_test.go#L133)):*
can also use `util.ErrorSkipFrames(1, "INSERT FOR ....", key, count)`

---

*[sql/txn_restart_test.go, line 267 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb67mPwnblJtScYwbr:-KCb67mPwnblJtScYwbs:626075299) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/txn_restart_test.go#L267)):*
`testutils.Is{E,Pe}rror`

---

*[sql/txn_restart_test.go, line 273 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb6ixCc6zoBT0VK5Bd:-KCb6ixDy0hc4fswZlwt:-1280018352) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/txn_restart_test.go#L273)):*
`WHERE true` isn't necessary.

---

*[sql/txn_restart_test.go, line 300 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb7b7-0Hsnql2RiSMB:-KCb7b70WVwFZzvaRrTT:-908825599) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/txn_restart_test.go#L300)):*
That's only an issue on write-write, do you want to push or abort? You're currently aborting, so the comment is misleading. If you want to push, SELECT below.

---

*[sql/txn_restart_test.go, line 383 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb8-DSQcO70EhgkvAI:-KCb8-DSQcO70EhgkvAJ:-1455889418) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/txn_restart_test.go#L383)):*
I thought that was already done somewhere else. In any case, drop `WHERE true`.

---

*[sql/txn_restart_test.go, line 412 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb8cgnDdDOfqJpBHtZ:-KCb8cgoQIU8_cyR-fSN:1676697132) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab02dd3b81c29699af1d9c7e576e934bc13a659f/sql/txn_restart_test.go#L412)):*
`testutils.IsError`

---

*[sql/txn_restart_test.go, line 418 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb8ehRiF3q02cuMJgn:-KCb8ehRiF3q02cuMJgo:1739356634) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab02dd3b81c29699af1d9c7e576e934bc13a659f/sql/txn_restart_test.go#L418)):*
ditto (and everywhere else I might've missed it)

---

*[sql/txn_restart_test.go, line 423 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb8s-Xq5nQ8JwwtRPo:-KCb8s-Xq5nQ8JwwtRPp:1875938707) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab02dd3b81c29699af1d9c7e576e934bc13a659f/sql/txn_restart_test.go#L423)):*
s/used/user/

---

*[sql/txn_restart_test.go, line 463 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb96FfoNIpvAJ5pf1y:-KCb96FfoNIpvAJ5pf1z:-1480476831) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab02dd3b81c29699af1d9c7e576e934bc13a659f/sql/txn_restart_test.go#L463)):*
where's this coming from? Copy pasta?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-:-KCb9D3aPbr9-3ASUuyL:-675599653)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
@tamird i've moved the tests to the last commit (and improved it now) because there was too much churn in there to make sense of it. We now inject errors in the COMMITs.

---

Review status: 17 of 23 files reviewed at latest revision, 36 unresolved discussions, some commit checks failed.

---

*[client/db.go, line 396 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCaxyeKfI3K2tPDZPYy:-KCr4Pup5xRUJPJDF2s3:1622997028) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/client/db.go#L396)):*
I respectfully disagree. It tripped me before, when it was like that.

---

*[client/txn.go, line 502 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCavyVVvr4s6aNPuX98:-KCr55zi6f4zpg5e1qaw:-1466866983) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ddd0104f832467d80279d38f9172a90eda45577/client/txn.go#L502)):*
in the first commit i've already made `CleanupOnError()` only callable when there's an error...

---

*[client/txn.go, line 454 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCayiiYheeKLkxED1Aw:-KCr4wLRTm5IhsO8IYiH:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/client/txn.go#L454)):*
Done.

---

*[roachpb/errors.go, line 559 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCXlPVX5M08gNR9YoxR:-KCr6KXzPB_5CymYFbqh:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/roachpb/errors.go#L559)):*
Done.

---

*[roachpb/errors.go, line 572 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCXlUgkRRKyfxsMYxxG:-KCr6X9xeaWLYGNqUKWa:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/roachpb/errors.go#L572)):*
Done.

---

*[roachpb/errors.proto, line 180 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCayyS3evosstw8khlc:-KCr6hAFWPJZsZ0Bnxwn:-870352293) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/roachpb/errors.proto#L180)):*
There's a TODO from @kaneda inside ErrorDetail to this effect.

---

*[roachpb/errors.proto, line 222 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCXlpq7-kDhcG9u0F-N:-KCr7191SLXKTv8wfg4s:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/roachpb/errors.proto#L222)):*
Done.

---

*[sql/executor.go, line 602 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCXhIUecFH0pCsTB2s4:-KCrDfFX3n0kp9ZEpz_E:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ddd0104f832467d80279d38f9172a90eda45577/sql/executor.go#L602)):*
Done.

---

*[sql/executor.go, line 889 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCXjHOteduJ1tzpfOun:-KCrUZpmQ_D16UzWDaRH:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ddd0104f832467d80279d38f9172a90eda45577/sql/executor.go#L889)):*
Done.

---

*[sql/executor.go, line 892 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCawhdOKR75yUI8g4SQ:-KCrGKnV8LxzLllUxt7O:-444810963) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ddd0104f832467d80279d38f9172a90eda45577/sql/executor.go#L892)):*
well technically the implicit txn is aborted at this point. But note that execRequest() resets everything for implicit txns after the `txn.Exec()` call

---

*[sql/executor.go, line 931 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCaxUbixj1ZTHtawUAP:-KCrHPAWAYsQzkQomvtY:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ddd0104f832467d80279d38f9172a90eda45577/sql/executor.go#L931)):*
Done.

---

*[sql/executor.go, line 932 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCXjq-PZrLSqxC4i-Xs:-KCrHliESBqrEvtLFUfb:1320506173) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ddd0104f832467d80279d38f9172a90eda45577/sql/executor.go#L932)):*
I think it makes sense as a method on the executor, by analogy with `CommitSQLTransaction` below. 
I've made the receiver unnamed to signify it's not being used.

---

*[sql/executor.go, line 411 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb1QfspuwCuULguKRy:-KCr7Zt7Uerm5GDuDQxX:-1828087825) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/executor.go#L411)):*
meh, I dunno. I think I prefer them separate... One is controlling txn.Exec...

---

*[sql/executor.go, line 422 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb-mLLWdHS5HcU0vhf:-KCrAyg97hXPEeIJ6iTF:-610795825) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/executor.go#L422)):*
added a little something, see if that makes it more clear... With sketchy field like this you're bound to have to look at the code that uses it if you care.

---

*[sql/executor.go, line 444 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb04pYKhExdioW_AFG:-KCrBS76179CdQ9BfRGB:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/executor.go#L444)):*
Done.

---

*[sql/executor.go, line 446 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb0QZf9qLHRIKIy1fD:-KCrChNImGEdMnj4Y6U9:1158308395) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/executor.go#L446)):*
added TODO

---

*[sql/executor.go, line 618 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb15zEsYa0e83BaKFv:-KCrD0SJS221dToNvaSf:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/executor.go#L618)):*
Done.

---

*[sql/executor.go, line 649 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb1s8OCO8WDysOgAa0:-KCrF0KbW03pLEz4CbjN:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/executor.go#L649)):*
Done.

---

*[sql/executor.go, line 745 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb2ObtrBbxQhuuol5Q:-KCrFKk1aZPtEUg7__jz:1226787040) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/executor.go#L745)):*
the caller is `txn.Exec`, which knows nothing about `txnState`...

---

*[sql/executor.go, line 853 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb2kDdaNt2wzqpJ0JS:-KCrFTdck7eXDY1YQJjN:792162456) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/executor.go#L853)):*
well that state can either be ABORTED (nothing is restarting) or restarting... We should arguably separate this function into two.

---

*[sql/executor.go, line 921 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb3PZprftmMkuxqUXE:-KCrFfqapPIhIRj_f59j:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/executor.go#L921)):*
Done.

---

*[sql/executor.go, line 955 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb3TtXQjkmQXdBmkqg:-KCrFhDc1PNsRZbYKH53:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/executor.go#L955)):*
Done.

---

*[sql/executor.go, line 987 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb3uQfdYvrsbuVn2Se:-KCrH17jGZCK-yjP5mXZ:-2118994929) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/executor.go#L987)):*
`RETRY INTENT` is only valid at the beginning of the txn, like `SET PRIORITY`. So in the RestartTransaction case, we can't be at the beginning of the txn.

---

*[sql/executor.go, line 1052 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb4H_a3q4PciBdIQkb:-KCrHruhoVbcJLIZl0vq:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/executor.go#L1052)):*
Done.

---

*[sql/session.proto, line 35 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCXm3Q35ZnRYTg8b2RP:-KCrIIXDD0-FoS0UTUsI:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/session.proto#L35)):*
Done.

---

*[sql/txn_restart_test.go, line 63 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb4vFias54nuOgI2fw:-KCrTZhU4mMKqn6NPoWM:190511693) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/txn_restart_test.go#L63)):*
... that tried to write the value `value`... :)
It's referencing an arg name

---

*[sql/txn_restart_test.go, line 267 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb67mPwnblJtScYwbr:-KCrSGA0IbKfu9JrbD8A:49624351) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/txn_restart_test.go#L267)):*
nice, thanks

---

*[sql/txn_restart_test.go, line 273 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb6ixCc6zoBT0VK5Bd:-KCrKi9zCcRTgrZjwUFn:1530545499) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/txn_restart_test.go#L273)):*
damn well should be

---

*[sql/txn_restart_test.go, line 300 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb7b7-0Hsnql2RiSMB:-KCrSKD_skfzWQRL9wB3:-473570831) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/txn_restart_test.go#L300)):*
right, changed to `abortTxn`. We inject retry errors in other places.

---

*[sql/txn_restart_test.go, line 383 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb8-DSQcO70EhgkvAI:-KCrIdB1HeZNmevt-0mw:969144637) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/txn_restart_test.go#L383)):*
I like `WHERE true`. In fact MySQL refuses to run deletes without a WHERE clause (at least in sane configurations).

---

*[sql/txn_restart_test.go, line 412 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb8cgnDdDOfqJpBHtZ:-KCrPraRcmBnwdiL-Jt1:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab02dd3b81c29699af1d9c7e576e934bc13a659f/sql/txn_restart_test.go#L412)):*
Done.

---

*[sql/txn_restart_test.go, line 418 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb8ehRiF3q02cuMJgn:-KCrSajjSmEDohL8phbH:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab02dd3b81c29699af1d9c7e576e934bc13a659f/sql/txn_restart_test.go#L418)):*
Done.

---

*[sql/txn_restart_test.go, line 423 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb8s-Xq5nQ8JwwtRPo:-KCrSij_zZOnIu5ptWTt:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab02dd3b81c29699af1d9c7e576e934bc13a659f/sql/txn_restart_test.go#L423)):*
Done.

---

*[sql/txn_restart_test.go, line 463 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb96FfoNIpvAJ5pf1y:-KCrSk3lxQBen4pscHzp:905541315) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab02dd3b81c29699af1d9c7e576e934bc13a659f/sql/txn_restart_test.go#L463)):*
yeah, that's the guy that aborted us

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-:-KCY3poHDsn3ipgJY9XR:519003150)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 17 of 23 files reviewed at latest revision, 37 unresolved discussions, some commit checks failed.

---

*[sql/parser/sql.y, line 1653 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCrbpb9pB3HVVo1D4s6:-KCrbpb9pB3HVVo1D4s7:952369395) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c2f91ffae9cd2f41b07186df182045db904cb64/sql/parser/sql.y#L1653)):*
Please add some test cases to `TestParse` and/or `TestParse2` for these new statements.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116)*
<!-- Sent from Reviewable.io -->

tamird commented:
Uh, OK. This looks worse, since now you have a commit with a message "review - improving test" - as far as future readers are concerned, this message may as well be empty.

---

Reviewed 4 of 6 files at r4.
Review status: 21 of 23 files reviewed at latest revision, 32 unresolved discussions, some commit checks failed.

---

*[sql/executor.go, line 932 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCXjq-PZrLSqxC4i-Xs:-KCubKYcnbgxgjoCgKdT:-2102331766) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ddd0104f832467d80279d38f9172a90eda45577/sql/executor.go#L932)):*
I'm going to say this again: this refactor doesn't make any sense.

---

*[sql/executor.go, line 454 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCuayRKpC6EWsNes1xY:-KCuayRLDrGS1QGtOr0u:166874) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c2f91ffae9cd2f41b07186df182045db904cb64/sql/executor.go#L454)):*
?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-:-KCua3o5mTuH7PGMKSPJ:-1141651329)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Needs a rebase. My comments are mostly addressed by I agree with @tamird on his.

---

Reviewed 6 of 6 files at r4.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[client/db.go, line 396 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCaxyeKfI3K2tPDZPYy:-KCuidTY6YVGtf7VQUNH:1879645522) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/client/db.go#L396)):*
It's fine to disagree, but what's the reason? Every call site manually verifies that the error is non-nil. I'm not going to press this, but I want to understand why.

---

*[sql/executor.go, line 411 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb1QfspuwCuULguKRy:-KCujSPTHEyphq0FhuUI:-499481822) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/executor.go#L411)):*
but isn't the way a txn executes tied to the txnState? It's nice to have things which are intermingled bundled up which gives a convenient place to document their interplay. I don't have enough of a bird-eye view over this diff or code in general, though.

---

*[sql/executor.go, line 745 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb2ObtrBbxQhuuol5Q:-KCuk5azQdELWekg4tMt:1360908692) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/executor.go#L745)):*
No, the caller is

```go
		txnClosure := func(txn *client.Txn, opt *client.TxnExecOptions) *roachpb.Error {
			return runTxnAttempt(e, planMaker, origState, txnState, txn, opt, stmtsToExec,
				&results, &remainingStmts)
		}

```

which should be able to run the assertion, set up `txnState` and then you can avoid passing `txn` into the closure (that's what I wanted above but wasn't what I wrote, sorry).

---

*[sql/executor.go, line 853 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb2kDdaNt2wzqpJ0JS:-KCul3vB4vlzYbkjWVt9:-555223677) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/executor.go#L853)):*
if it's aborted, wouldn't the client be expected to restart (in which case the name `RestartingTxn` sounds better than the status quo)?

---

*[sql/txn_restart_test.go, line 300 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb7b7-0Hsnql2RiSMB:-KCumQ5onWKF89DCKAWY:908785948) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/txn_restart_test.go#L300)):*
What guarantees that the priorities involved allow a push? I'd rather not have more tests that sit around doing nothing for most of the time.

---

*[sql/txn_restart_test.go, line 383 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb8-DSQcO70EhgkvAI:-KCumlNhlvu2DjsKuVks:542291720) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/txn_restart_test.go#L383)):*
Ok, why not.

---

*[sql/txn_restart_test.go, line 463 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb96FfoNIpvAJ5pf1y:-KCumyq1GzvV8mHleDHg:1176579693) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab02dd3b81c29699af1d9c7e576e934bc13a659f/sql/txn_restart_test.go#L463)):*
It says above that we "don't see any rows". Why don't you let that aborting write be followed by a delete (in the same txn) so that there's no weird key around when we check here?

---

*[sql/txn_restart_test.go, line 165 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCulzH5GkYJxpBWHvKz:-KCulzH5GkYJxpBWHvL-:471535059) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c2f91ffae9cd2f41b07186df182045db904cb64/sql/txn_restart_test.go#L165)):*
nit: `var fail bool`

---

*[sql/txn_restart_test.go, line 195 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCum80rqfubXQnhfq3h:-KCum80rqfubXQnhfq3i:1962307351) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c2f91ffae9cd2f41b07186df182045db904cb64/sql/txn_restart_test.go#L195)):*
I think you can ditch `fail` and just use `t.Failed()` instead.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-:-KCunGmJS3KfMbsFLQai:-1194265034)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Posting comments for the first commit, sorry if some of them are on stale code.

---

Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[client/txn.go, line 345 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCuyaMeKpfS38IMTPZ-:-KCuyaMeKpfS38IMTPZ0:-2042079705) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ddd0104f832467d80279d38f9172a90eda45577/client/txn.go#L345)):*
function name will be obvious in the trace, can just say "no error"

---

*[client/txn.go, line 502 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCuysUdgCKYqOkxFJxI:-KCuysUdgCKYqOkxFJxJ:1985015873) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ddd0104f832467d80279d38f9172a90eda45577/client/txn.go#L502)):*
cleanup or CleanupOnError

---

*[sql/executor.go, line 932 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCXjq-PZrLSqxC4i-Xs:-KCv-EdLTG7H1qwY_sE7:1793347520) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ddd0104f832467d80279d38f9172a90eda45577/sql/executor.go#L932)):*
If you want to keep them here, I would make them "bare" functions instead of Executor methods.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-:-KCv-_IvobUDAzbi35HF:507897174)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
You can also add a partest (see the existing one in sql/partestdata) to verify that parallel conflicting transactions that use RETRY INTENT eventually succeed.

---

Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[client/db.go, line 396 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCaxyeKfI3K2tPDZPYy:-KCv08HbxcLpluQE_aph:-880546942) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/client/db.go#L396)):*
I can understand why it may be helpful when reading the code to know that the call doesn't do anything if there is no error. The "if" makes this obvious. If we want to do it in the function, a better method name might help, e.g. `CleanupIfError`

---

*[sql/executor.go, line 438 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCv1VphtU4vZ7jaA9Xf:-KCv1VpiTtQ7EgWj99Nv:1777598005) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c2f91ffae9cd2f41b07186df182045db904cb64/sql/executor.go#L438)):*
I think this function deserves a comment

---

*[sql/executor.go, line 459 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCv1mgpu7RN9csSmWuc:-KCv1mgpu7RN9csSmWud:274402227) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c2f91ffae9cd2f41b07186df182045db904cb64/sql/executor.go#L459)):*
it would be clearer if this was under the case IMMEDIATE above

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-:-KCv48BKqaPdpd1d9pNG:242532503)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
@tamird I'll squash that commit (and all the review commits if there'll be more) into the last one 2nd one ("Support user-directed retries...")

---

Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[client/db.go, line 396 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCaxyeKfI3K2tPDZPYy:-KD-tI5tb1zc8RReUO6I:-1817442794) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/client/db.go#L396)):*
yeah, what Radu said. I prefer it this way since in the latest code, this function is called unconditionally from blocks that are already handling errors.

---

*[client/txn.go, line 345 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCuyaMeKpfS38IMTPZ-:-KD-uOFsDohGveMbd89t:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ddd0104f832467d80279d38f9172a90eda45577/client/txn.go#L345)):*
Done.

---

*[client/txn.go, line 502 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCuysUdgCKYqOkxFJxI:-KD-uavVcNRtPd6uJuB5:721678128) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ddd0104f832467d80279d38f9172a90eda45577/client/txn.go#L502)):*
this code is gone (TODO is addressed) in next commit

---

*[sql/executor.go, line 932 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCXjq-PZrLSqxC4i-Xs:-KD05VVBxg9maPLF5HGQ:-816254040) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ddd0104f832467d80279d38f9172a90eda45577/sql/executor.go#L932)):*
ok, bare it is. This PR should have probably contained a separate commit for moving `{commit,rollback}SQLTransaction` here from the planner. The point is that now the planner doesn't know anything about committing or otherwise affecting the state of a txn because the way it signaled back to the executor that it did so was awkward.

---

*[sql/executor.go, line 411 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb1QfspuwCuULguKRy:-KD-v2L6Z9OWjgtM-zwY:-1758742251) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/executor.go#L411)):*
well they're somewhat related... but they still feel separate to me.

---

*[sql/executor.go, line 745 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb2ObtrBbxQhuuol5Q:-KD01A4hVFhuUmmqBWBS:1660673450) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/executor.go#L745)):*
yup good call. Done.

---

*[sql/executor.go, line 853 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb2kDdaNt2wzqpJ0JS:-KD01ioPkWCITCzCb0pi:1288002275) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/executor.go#L853)):*
no, if the txn is in state `ABORTED` there will be no restart (either because the error we got is not retriable, or because the we're neither in an auto-retry txn nor in a `RETRY INTENT` txn). The KV txn has been rolled back.
We go into state `RESTART_WAIT` if we're going to be retried.

---

*[sql/executor.go, line 438 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCv1VphtU4vZ7jaA9Xf:-KD-w9B9DhpBioilB6qc:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c2f91ffae9cd2f41b07186df182045db904cb64/sql/executor.go#L438)):*
Done.

---

*[sql/executor.go, line 454 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCuayRKpC6EWsNes1xY:-KD0-pmqznYwNOEhUWG_:-172567796) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c2f91ffae9cd2f41b07186df182045db904cb64/sql/executor.go#L454)):*
took Radu's suggestion below, now the `fallthrough` actually does smth

---

*[sql/executor.go, line 459 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCv1mgpu7RN9csSmWuc:-KD0-wDvljp5SJfKPk87:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c2f91ffae9cd2f41b07186df182045db904cb64/sql/executor.go#L459)):*
Done.

---

*[sql/txn_restart_test.go, line 300 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb7b7-0Hsnql2RiSMB:-KD081pamxHxQxQGSrT9:-1186711756) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/txn_restart_test.go#L300)):*
well no guarantees, it's reliable because of the spinning, as the comment below says. But Radu's introducing syntax for deterministic priorities, added TODOs to do that. In the meantime I've stressed this test, works nicely.

---

*[sql/txn_restart_test.go, line 383 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb8-DSQcO70EhgkvAI:-KD07Cpy0XOo32wVU0UF:-367785279) ([raw file](https://github.com/cockroachdb/cockroach/blob/cab6ebefd8be7e050422387ee30d927795d2b71d/sql/txn_restart_test.go#L383)):*
because it's way too easy to delete everything in a table, which is rarely what you want

---

*[sql/txn_restart_test.go, line 463 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCb96FfoNIpvAJ5pf1y:-KD0BA6wJpUTrPKgSs24:-1465647342) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab02dd3b81c29699af1d9c7e576e934bc13a659f/sql/txn_restart_test.go#L463)):*
ok, changed the aborting guy to just a delete. Boris, we hardly knew ye.

---

*[sql/txn_restart_test.go, line 195 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCum80rqfubXQnhfq3h:-KD06IHIUoy6AFnJZtnO:-1952617514) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c2f91ffae9cd2f41b07186df182045db904cb64/sql/txn_restart_test.go#L195)):*
nice, done.

---

*[sql/parser/sql.y, line 1653 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KCrbpb9pB3HVVo1D4s6:-KD0C33PVzPakwXPKT7S:-1299657660) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c2f91ffae9cd2f41b07186df182045db904cb64/sql/parser/sql.y#L1653)):*
done, thanks

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-:-KD-soGoRWe1naQYd2dB:-2084484000)*
<!-- Sent from Reviewable.io -->

andreimatei committed:
sql: add a test for RESTART after a non-retriable error

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 15 of 24 files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[roachpb/errors.go, line 572 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KD0UgPPD_fOcDl-FbRb:-KD0UgPPD_fOcDl-FbRc:1079677152) ([raw file](https://github.com/cockroachdb/cockroach/blob/10c60e0fe8c33429683e1504667d3f29f9d27c6a/roachpb/errors.go#L572)):*
why are these two functions not similar? We don't need CustomMsg in SqlTransactionCommittedError?

---

*[sql/executor.go, line 634 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KD0VEb21m4ap9pjCZye:-KD0VEb21m4ap9pjCZyf:-1714772490) ([raw file](https://github.com/cockroachdb/cockroach/blob/10c60e0fe8c33429683e1504667d3f29f9d27c6a/sql/executor.go#L634)):*
We could just call SetDebugName in each branch - that functions remembers the line where it was called from so it might make things a bit easier when debugging

---

*[sql/executor.go, line 1062 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KD0XcbgYGOeg9FR6CKG:-KD0XcbhA0UV-_gz3Mo8:-859217721) ([raw file](https://github.com/cockroachdb/cockroach/blob/10c60e0fe8c33429683e1504667d3f29f9d27c6a/sql/executor.go#L1062)):*
We should mention the "release" functionality (which depends on stmt)

---

*[sql/executor.go, line 1064 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KD0WBLfjWmRs3dSP34L:-KD0WBLfjWmRs3dSP34M:1990053115) ([raw file](https://github.com/cockroachdb/cockroach/blob/10c60e0fe8c33429683e1504667d3f29f9d27c6a/sql/executor.go#L1064)):*
maybe we can pass just whether it is a release or commit as an argument instead of stmt (the calling code also looks at stmt so the logic is a bit confusing)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-:-KD0_9b40kE2rRS-hbI3:1417088721)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 1 of 24 files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[roachpb/errors.go, line 572 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KD0UgPPD_fOcDl-FbRb:-KD9YluFBrGF03la0-_m:-271874806) ([raw file](https://github.com/cockroachdb/cockroach/blob/10c60e0fe8c33429683e1504667d3f29f9d27c6a/roachpb/errors.go#L572)):*
correct, the committed error doesn't need a custom msg. We only use the custom msg in a special case when the transaction is in the aborted case.

---

*[sql/executor.go, line 634 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KD0VEb21m4ap9pjCZye:-KD9qXc_r3xzXg8dWqaP:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/10c60e0fe8c33429683e1504667d3f29f9d27c6a/sql/executor.go#L634)):*
Done.

---

*[sql/executor.go, line 1062 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KD0XcbgYGOeg9FR6CKG:-KD9qtdYiwcpJmlQm9_s:-2116073067) ([raw file](https://github.com/cockroachdb/cockroach/blob/10c60e0fe8c33429683e1504667d3f29f9d27c6a/sql/executor.go#L1062)):*
i've added an argument specifying the commit type, as per the suggestion below. release functionality is now more visible.

---

*[sql/executor.go, line 1064 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116#-KD0WBLfjWmRs3dSP34L:-KD9tlqHXxgTcmK5z8kC:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/10c60e0fe8c33429683e1504667d3f29f9d27c6a/sql/executor.go#L1064)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Reviewed 5 of 5 files at r1.
Review status: 1 of 24 files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5116)*
<!-- Sent from Reviewable.io -->
